Fix for issue #9: wallet not staking after unlocking

fopat opened the pull request:
## Issue
When using the "walletpassphrase" command in the debug console or with the daemon, the wallet did not start staking if a value too large was entered as timeout.

## Cause
The timeout parsing and management is in rpcwallet.cpp, in the walletpassphrase function.
There are 2 limits for this timeout value:
- Entering a value larger than the max int64 value will cause the nSleepTime variable to be negative
- Entering a value larger than 999999000000000 (approximatively) will result in a negative value when calling posix_time::seconds(nSeconds) (rpcserver.cpp:665)

In both cases, the timeout becomes negative and the wallet is immediatly relocked, so the staking is aborted.

## Fix proposal
IMHO, there is no easy way to accept larger values than 999999000000000, so I added a test to handle both cases and to return an error message informing that the timeout value is incorrect.

Matthematic commented on src/rpcwallet.cpp:
```diff
@@ -1963,6 +1963,14 @@ Value walletpassphrase(const Array& params, bool fHelp)
+    // Timeout limits:
+    // If the timeout is larger than the max value of int64 (922337203685477580), nSleepTime will be negative
+    // If the timeout is larger than 999999000000000, boost will not be able to convert it to seconds
+    if (nSleepTime < 0 || nSleepTime > 999999000000000)
```
Can we throw the warning before even converting by just comparing it to the max of int64_t? Also I think the comment of 922337203685477580 is missing the last digit - I think its 9223372036854775807.

`
#include <limits>
if (nSleepTime > std::numeric_limits<int64_t>::max()) { throw JSONRPCError(); }
`

It would be best if we could find the actual type that posix_time::seconds() uses, as it could possibly change based on compiler, architecture, etc. Best I could find was that it was a long from http://www.boost.org/doc/libs/1_55_0/doc/html/date_time/posix_time.html#date_time.posix_time.time_duration but I'm not confident if that's the same function. How did you come up with an approx value of 999999000000000?

fopat replied:
You are right on both points:
- Finding a compiler-independant fix would be much better
- I forgot the last digit in the comment ^^

I found the value with some tests, and tried to find the exact type but I have not managed yet.
However, I have just had an idea of a way to fix it based on a Boost type.
I will submit a PR update shortly.

fopat commented:
I have just updated the fix to a solution that should work on any compiler, whatever types Boost is using.
By trying to convert the timeout value to the boost::posix_time::second type, we can ensure that this timeout will never be negative or null.
Also, I added a call to CWallet::lock to relock the wallet if the unlock operation fails because of an incorrect timeout value.

What do you think?

fopat commented:
On a 64-bit Linux compiler, the long type seems to be 64-bits and not 32.
So std::numeric_limits<long>::max() is the max value of an int64, which is 9223372036854775807.

However, according to my tests, posix_time::seconds becomes negative after approximately 999999000000000, which is far below the max long value.
The entry type of posix_time::seconds is indeed a long, but it gets converted with a static_cast to the sec_type type (posix_time_duration.hpp:46, in the constructor of the seconds class).
I have tried to figure out the exact type of sec_type, but it seems to be dependant of the time resolution.
Also, later in the walletpassphrase function, the timeout is added to the current time, so GetTime() + nSleepTime shall not exceed the max int64 value.

The easiest solution would indeed to limit the timeout to INT32_MAX (2147483647). As you said, it would allow more than 50 years and would be the most efficient solution.

Should I update the PR with this solution?

dragonsmilk commented:
Go ahead and limit it to INT32_MAX so we aren't playing with fire when cross-compiling across platforms.

TechniumUnlimited commented:
Merging to dev
